scripts/builtin/shortestPath.dml

```diff
@@ -0,0 +1,157 @@
+#	will be saved in this matrix.
+#
+#
+#
+
+
+s_shortestPath = function(Matrix[Double] G) 
+  return (Matrix[Double] C) 
+{
+	
+	print("SHORTEST PATH CALCULATION");
+
+	if(min(G) < 0){
+		stop("All values in G must be positive")
+	}
+	
+	if(ncol(G) != nrow(G)){
+		stop("Not correct matrix dimensions")
+	}
+	
+	matrixSize = nrow(G)
+	infValue = sum(rowSums(G)) + 1 # value representing infinity, i.e. the nodes are not connected
+	
+	# initialize the matrix of minimum distances with "infinity" values:
+	minDistMatrix = matrix(infValue,rows=matrixSize,cols=matrixSize)
+	
+	for (sourceNode in 1:matrixSize){ 
+		
+		# update minimum distance from the sourceNode to itself to 0:
+		minDistMatrix[sourceNode,sourceNode] = 0
+		
+		# initialize the matrix neighboursList and neighboursListTmp:
+		neighboursList  = matrix(0,matrixSize,3)
+		neighboursListTmp = matrix(0,matrixSize*matrixSize,3)
+		
+		# find the neighbours of the sourceNode and fill in the neighboursList:
+		nodeIdx = 1
+		for(ineighbour in 1:matrixSize){
+			if (as.integer(as.scalar(G[sourceNode,ineighbour]))>0 & (ineighbour!=sourceNode)){ 
+				neighboursList[nodeIdx,1] = sourceNode
+				neighboursList[nodeIdx,2] = ineighbour
+				nodeIdx = nodeIdx+1
+			}
+		}
+		
+		
+		while( as.integer(as.scalar(neighboursList[1,1])) > 0 ){ # loop of supersteps (see documentation)
+			rawIdx = 1
+			nodeIdx=1		
+			
+			while( as.integer(as.scalar(neighboursList[rawIdx,1])) > 0 ){ # loop over the raws of neighboursList, i.e. updates of the previous superstep
+				
+				# define the currentNode, previousNode and cumulativeMinimumDistance from the corresponding raw of neighboursList:
+				currentNode = as.integer(as.scalar(neighboursList[rawIdx,2])) 
+				previousNode = as.integer(as.scalar(neighboursList[rawIdx,1])) 
+				cumulativeMinimumDistance = as.integer(as.scalar(neighboursList[rawIdx,3]))
+				
+				# consider if the minimum distance from the source node to the current node can be updated:
+				potentialMinimumDistance = cumulativeMinimumDistance + as.integer(as.scalar(G[previousNode,currentNode]))
+				if (potentialMinimumDistance < as.integer(as.scalar(minDistMatrix[sourceNode,currentNode]))){
+					minDistMatrix[sourceNode,currentNode] = potentialMinimumDistance
+	
+					# find the neighbours of currentNode to send them the updated information:
+					for(ineighbour in 1:matrixSize){
+						if (as.integer(as.scalar(G[currentNode,ineighbour]))>0 & (ineighbour!=sourceNode)){ 
+							neighboursListTmp[nodeIdx,1] = currentNode
+							neighboursListTmp[nodeIdx,2] = ineighbour
+							neighboursListTmp[nodeIdx,3] = potentialMinimumDistance
+							nodeIdx = nodeIdx+1
+						}
+					}
+				}
+				rawIdx=rawIdx+1
+			}
+			
+			# update neighboursList and neighboursListTmp for the next superstep:
+			neighboursList = neighboursListTmp
+			neighboursListTmp = matrix(0,matrixSize*matrixSize,3)
+		}	
+	}
+	
+	# define the distance between the not connected nodes as -1:
+	
+	for (irow in 1:matrixSize){
```

Comment:
can be dropped - simply initialize the vector of min distances as inf and if a node is not reachable this entry will never be updated.